packages/graph/hash_graph/lib/graph/src/types/schema/reference.rs

```diff
@@ -0,0 +1,143 @@
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize, sqlx::Type)]
+#[serde(transparent)]
+#[sqlx(transparent)]
+pub struct Uri(String);
+
+impl Uri {
+    /// Creates a new `Uri` without validating.
+    #[must_use]
+    pub fn new_unchecked<T: Into<String>>(uri: T) -> Self {
+        Self(uri.into())
+    }
+
+    /// Creates a new `Uri` from the given string.
+    ///
+    /// # Errors
+    ///
+    /// - [`ValidationError`] if validation is failing.
+    pub fn new<T: Into<String>>(uri: T) -> Result<Self, ValidationError> {
+        let uri = Self::new_unchecked(uri);
+        uri.validate()?;
+        Ok(uri)
+    }
+}
+
+impl fmt::Display for Uri {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", self.0)
+    }
+}
+
+impl Validate for Uri {
+    fn validate(&self) -> Result<(), ValidationError> {
+        Ok(())
+    }
+}
+
+/// Property Object values must be defined through references to the same valid URI to a Property
+/// Type.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct PropertyTypeReference {
+    #[serde(rename = "$ref")]
+    reference: Uri,
+}
+
+impl PropertyTypeReference {
+    /// Creates a new `PropertyTypeReference` without validating.
+    #[must_use]
+    pub const fn new_unchecked(reference: Uri) -> Self {
+        Self { reference }
+    }
+
+    /// Creates a new `PropertyTypeReference` from the given `reference`.
+    ///
+    /// # Errors
+    ///
+    /// - [`ValidationError`] if validation is failing.
+    pub fn new(reference: Uri) -> Result<Self, ValidationError> {
+        let reference = Self::new_unchecked(reference);
+        reference.validate()?;
+        Ok(reference)
+    }
+
+    #[must_use]
+    pub const fn reference(&self) -> &Uri {
+        &self.reference
+    }
+}
+
+impl Validate for PropertyTypeReference {
+    fn validate(&self) -> Result<(), ValidationError> {
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct DataTypeReference {
+    #[serde(rename = "$ref")]
+    reference: Uri,
+}
+
+impl DataTypeReference {
+    /// Creates a new `DataTypeReference` without validating.
+    #[must_use]
+    pub const fn new_unchecked(reference: Uri) -> Self {
+        Self { reference }
+    }
+
+    /// Creates a new `DataTypeReference` from the given `reference`.
+    ///
+    /// # Errors
+    ///
+    /// - [`ValidationError`] if validation is failing.
+    pub fn new(reference: Uri) -> Result<Self, ValidationError> {
+        let reference = Self::new_unchecked(reference);
+        reference.validate()?;
+        Ok(reference)
+    }
+
+    #[must_use]
+    pub const fn reference(&self) -> &Uri {
+        &self.reference
+    }
+}
+
+impl Validate for DataTypeReference {
+    fn validate(&self) -> Result<(), ValidationError> {
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
nit:
Should we also have a `property_type_reference` test here for consistency?

Reply:
Added as part of d2fbcfecbd8ae0b5235b5e3ffdc96688edbc11cc